packages/core/src/modules/connections/ConnectionsModule.ts

```diff
@@ -225,37 +220,28 @@ export class ConnectionsModule {
     return this.connectionService.deleteById(connectionId)
   }
 
-  /**
-   * Find connection by verkey.
-   *
-   * @param verkey the verkey to search for
-   * @returns the connection record, or null if not found
-   * @throws {RecordDuplicateError} if multiple connections are found for the given verkey
-   */
-  public findByVerkey(verkey: string): Promise<ConnectionRecord | null> {
-    return this.connectionService.findByVerkey(verkey)
-  }
+  public async findByKeys({ senderKey, recipientKey }: { senderKey: string; recipientKey: string }) {
+    const theirDidRecord = await this.didRepository.findByRecipientKey(senderKey)
+    if (theirDidRecord) {
+      const ourDidRecord = await this.didRepository.findByRecipientKey(recipientKey)
+      if (ourDidRecord) {
+        const connectionRecord = await this.connectionService.findSingleByQuery({
+          did: ourDidRecord.id,
+          theirDid: theirDidRecord.id,
+        })
+        if (connectionRecord && connectionRecord.isReady) return connectionRecord
+      }
+    }
 
-  /**
-   * Find connection by their verkey.
-   *
-   * @param verkey the verkey to search for
-   * @returns the connection record, or null if not found
-   * @throws {RecordDuplicateError} if multiple connections are found for the given verkey
-   */
-  public findByTheirKey(verkey: string): Promise<ConnectionRecord | null> {
-    return this.connectionService.findByTheirKey(verkey)
+    this.agentConfig.logger.debug(
+      `No connection record found for encrypted message with recipient key ${recipientKey} and sender key ${senderKey}`
+    )
+
+    return null
```

Comment:
I agree that usually, the return early strategy is better. I reconsidered it but in this case, it would mean calling a return many times, which I also don't like, and returning null, which is even worse.